torch/csrc/distributed/c10d/Utils.cu

```diff
@@ -0,0 +1,27 @@
+#include <torch/csrc/distributed/c10d/Utils.hpp>
+
+#include <ATen/Dispatch.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+
+namespace c10d {
+
+__global__ void checkForNaN(float* data, int size) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx < size) {
+        float val = data[idx];
+        // Use device-side assert to check for NaN
+        // This will cause the kernel to terminate if NaN is found
+        CUDA_KERNEL_ASSERT(!isnan(val));
+    }
+}
+
+void checkForNan(const at::Tensor& tensor) {
+  const int blockSize = 256;
+  const int numBlocks = (tensor.numel() + blockSize - 1) / blockSize;
```

Comment:
(Related to my comment above)
Oh, I see, you are launching many blocks here. 
No sure if that's a good idea. CUDA has a limit on how many blocks you can launch. If the tensor is big, the next line will error out.

Reply:
I see, should be straightforward to fix the numBlocks to 1

Reply:
You can do some tuning based on `tensor.numel()`. Or, apply a max(a, b) to the current numBlocks line. 
Re blockSize, I think raising it to 512 would not be a crazy idea.